firstapp/app/src/androidTest/java/com/example/firstapp/ShellUtility.java

```diff
@@ -287,29 +330,151 @@ public Action parseStringAction(String str, int idx) throws invalidInputExceptio
         return res;
     }
 
-
     /**
      * Executes CUJ and returns cached objects
      */
-    public UiObject[] cacheCUJ(Action[] cuj) throws InterruptedException, invalidInputException, UiObjectNotFoundException, IOException {
-        UiObject[] cachedObjects = new UiObject[cuj.length];
+    public void  cacheCUJ(Action[] cuj) throws InterruptedException, invalidInputException, UiObjectNotFoundException, IOException {
         for (int i = 0; i < cuj.length; i++) {
-            cachedObjects[i] = cuj[i].executeUncachedAction();
+            cuj[i].executeUncachedAction();
         }
-        return cachedObjects;
     }
 
+
+                                /****************************
+                                 * DATA MANIPULATION HELPERS
+                                 ****************************/
+    public long getTime() {
+        return SystemClock.currentGnssTimeClock().millis();
+    }
+
+    public long sumArr(long[] arr) {
+        long total = 0;
+        for (long t : arr) total += t;
+        return total;
+    }
+
+    /**
+     * differences between successive elements
+     */
+    public long[] differences(long[] arr) {
+        if (arr.length == 0) return arr;
+        long[] res = new long[arr.length - 1];
+        for (int i = 0; i < res.length; i++) {
+            res[i] = arr[i + 1] - arr[i];
+        }
+        return res;
+    }
     /**
-     * Executes a CUJ with preparatory actions pre and measured actions post. Caches information for later use.
+     * value of last (n - 1) scaled, where first value is set to 0
      */
-    public void executeCUJ(String[] preCUJ, String[] postCUJ) throws Exception {
+    public long[] relativeValues(long[] arr) {
+        if (arr.length == 0) return null;
+        long[] res = new long[arr.length - 1];
+        for (int i = 0; i < res.length; i++) {
+            res[i] = arr[i + 1] - arr[0];
+        }
+        return res;
+    }
+
+    public long[] averageColumns(long[][] arr) {
+        if (arr == null || arr.length == 0) return null;
+        long[] averages = new long[arr[0].length];
+        for (int j = 0; j < arr[0].length; j++) {
+            long sum = 0;
+            for (int i = 0; i < arr.length; i++) {
+                sum += arr[i][j];
+            }
+            averages[j] = sum / arr.length;
+        }
+        return averages;
+    }
+
+    public String zeroPad(String s, int finalLength) {
+        while (s.length() < finalLength) {
+            s = "0" + s;
+        }
+        return s;
+    }
+
+    public void logData(long[][] allActionStamps) {
```

Comment:
Need to formally define the input and output of your script.  Will it be logging to logcat and forcing the user to scan logcat?  This should be written down in the design doc.

Reply:
script will redirect a summary to a file and display it on the screen